src/clients/node/src/node.zig

```diff
@@ -15,783 +19,465 @@ const CreateTransfersResult = tb.CreateTransfersResult;
+        .lookup_transfers => try decode_array_into_data(u128, Transfer, env, array),
+    };
 
-    const context = Context.create(env, allocator, &globals.io, cluster, addresses) catch {
-        // TODO: switch on err and provide more detailed messages
-        translate.throw(env, "Failed to initialize Client.") catch return null;
+    packet.* = .{
+        .next = null,
+        .user_data = callback_ref,
+        .operation = @intFromEnum(op),
+        .status = .ok,
+        .data_size = @intCast(data.len),
+        .data = data.ptr,
     };
 
-    return context;
+    tb.submit(client, packet);
 }
 
-/// This function decodes and validates an array of Node objects, one-by-one, directly into an
-/// available message before requesting the client to send it.
-fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
-    var argc: usize = 4;
-    var argv: [4]c.napi_value = undefined;
-    if (c.napi_get_cb_info(env, info, &argc, &argv, null, null) != c.napi_ok) {
-        translate.throw(env, "Failed to get args.") catch return null;
+fn on_completion(
+    on_completion_ctx: usize,
+    client: tb.tb_client_t,
+    packet: *tb.tb_packet_t,
+    result_ptr: ?[*]const u8,
+    result_len: u32,
+) callconv(.C) void {
+    switch (packet.status) {
+        .ok => {},
+        .too_much_data => unreachable, // We limit packet data size during request().
+        .invalid_operation => unreachable, // We check the operation during request().
+        .invalid_data_size => unreachable, // We set correct data size during request().
     }
 
-    if (argc != 4) translate.throw(
-        env,
-        "Function request() requires 4 arguments exactly.",
-    ) catch return null;
+    // Copy the results given to use into the packet dataa.
+    const events = @as([*]align(16) u8, @ptrCast(@alignCast(packet.data.?)))[0..packet.data_size];
+    const results = result_ptr.?[0..result_len];
+    const data_size = switch (@as(Operation, @enumFromInt(packet.operation))) {
+        .create_accounts => encode_results_into_data(Account, CreateAccountsResult, results, events),
+        .create_transfers => encode_results_into_data(Transfer, CreateTransfersResult, results, events),
+        .lookup_accounts => encode_results_into_data(u128, Account, results, events),
+        .lookup_transfers => encode_results_into_data(u128, Transfer, results, events),
+    };
 
-    const context_raw = translate.value_external(
-        env,
-        argv[0],
-        "Failed to get Client Context pointer.",
-    ) catch return null;
-    const context = contextCast(context_raw.?) catch return null;
-    const operation_int = translate.u32_from_value(env, argv[1], "operation") catch return null;
+    // Stuff the event size along with the packet result size into the packet.
+    // The former is needed to know the full heap allocation size in packet.data.
+    // The latter is needed to know how many Results in packet.data to serialize for JS.
```

Comment:
For a given op, packet memory is allocated to hold `max(sizeof(Event/Request) * n, sizeof(Result/Response))`. As long as the op and `n` are known, the allocation size can be computed.

When submitting the packet, the data_size is `sizeof(Event) * n`. On packet result, we overwrite the memory with `sizeof(Result) * completed` where completed <= n. If less Results are completed than Events, the completed size no longer reflects "n". So instead, data_size stores `packed(n, completed)`: arg1 to know alloc size, arg2 to know what to parse back to JS. 

Admittedly got a bit too "clever" here to avoid allocations.

Reply:
> Admittedly got a bit too "clever" here to avoid allocations.

😎

You should document it here, or at `DataSize`.
Btw, what about renaming `DataSize` to `BufferSize` or something else communicating that?